// nodeash/dist/cjs/server.connection.js
'use strict';
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const http_1 = __importDefault(require("http"));
const express_1 = __importDefault(require("express"));
const body_parser_1 = __importDefault(require("body-parser"));
const helmet_1 = __importDefault(require("helmet"));
const cors_1 = __importDefault(require("cors"));
const debug_1 = __importDefault(require("debug"));
const debugInstance = (0, debug_1.default)('node');
class ServerClient {
    constructor(options = {}) {
        this.initMiddleware = () => {
            this.app.set('port', this.port);
            this.app.use((0, helmet_1.default)());
            this.app.use((0, cors_1.default)());
            this.app.use(body_parser_1.default.json());
            this.app.use(body_parser_1.default.urlencoded({ extended: true, limit: '5mb' }));
            this.app.use((req, _res, next) => {
                console.log(`[${new Date().toISOString()}] ðŸ§ª ${req.method} ${req.url}`);
                next();
            });
        };
        this.initControllers = (controllers) => {
            if (controllers.length > 0) {
                controllers.forEach((controller) => {
                    this.app.use(`${this.baseUrl}`, controller.router);
                });
            }
            this.app.use(`${this.baseUrl}/status`, (_req, res) => {
                return res.send({
                    isSuccess: true,
                    memory: process.memoryUsage(),
                    results: {
                        message: `App is running on Port ${this.port}.`,
                    },
                    Date: new Date(),
                });
            });
        };
        this.startServer = () => {
            this.app_Server.on('error', this.onError);
            this.app_Server.on('listening', this.onListening);
            this.app_Server.listen(this.port);
        };
        this.onListening = () => {
            const bind = `port ${this.port}`;
            debugInstance('Listening on ' + bind);
            console.log(`á´€á´˜á´˜ Éªêœ± ÊŸÉªêœ±á´›á´‡É´ÉªÉ´É¢ á´É´ á´˜á´Ê€á´› ${this.port}`);
        };
        this.onError = (error) => {
            if (error instanceof Error) {
                throw error;
            }
            const errnoError = error;
            if (errnoError.syscall !== undefined) {
                const bind = `port ${this.port}`;
                switch (errnoError.code) {
                    case 'EACCES':
                        console.log(bind + ' requires elevated privileges');
                        break;
                    case 'EADDRINUSE':
                        console.log(bind + ' is already in use');
                        break;
                    default:
                        throw error;
                }
                process.exit(1);
            }
        };
        const { controllers = [], port = 3000, baseUrl = "" } = options;
        this.app = (0, express_1.default)();
        this.app_Server = http_1.default.createServer(this.app);
        this.port = port;
        this.baseUrl = baseUrl;
        this.initMiddleware();
        this.initControllers(controllers);
        // this.startProcess();
    }
}
exports.default = ServerClient;
